### PR TITLE
QEA-1213: Error checks to Gridium-TestRail

### PR DESCRIPTION
**Pull request by sethuster**
Want make sure that a failed connection to TestRail doesn't fail tests in Jenkins.  

Try to connect multiple times, if failed after multiple attempts - don't crash the tests.
If unable to create a test run, don't attempt to post results to a non-existent testrun.

Reduce the amount of API calls needed to create runs and update results.

### CODE REVIEW

File: lib/testrail.rb [hunk 1]
```diff
@@ -7,79 +7,102 @@ module Gridium
...
-
 		def initialize
       if Gridium.config.testrail
         @url = ENV['GRIDIUM_TR_URL'].empty? || ENV['GRIDIUM_TR_URL'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_URL'] + '/index.php?/api/v2/'
         @user = ENV['GRIDIUM_TR_USER'].empty? || ENV['GRIDIUM_TR_USER'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_USER']
         @password = ENV['GRIDIUM_TR_PW'].empty? || ENV['GRIDIUM_TR_PW'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_PW']
         @pid = ENV['GRIDIUM_TR_PID'].empty? || ENV['GRIDIUM_TR_PID'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_PID']
-				@testcase_ids = Array.new
-				@testcase_infos = Array.new
+				@config = {:pass => 1, :blocked => 2, :untested => 3, :retest => 4, :fail => 5}
```
Comment (rizzza) [inline lib/testrail.rb]: Nice

File: lib/testrail.rb [hunk 2]
```diff
@@ -7,79 +7,102 @@ module Gridium
...
       if Gridium.config.testrail
         Log.debug("[GRIDIUM::TestRail] Creating Test Run: name: #{name} desc: #{desc}")
         if name.nil? || name.empty? then
-          raise(ArgumentError, "Empty Run Name - Run name is required")
-        end
-        r = _send_request('POST', "add_run/#{@pid}", {:name => name, :description => desc, :include_all => false})
-        Log.debug("Result: #{r}")
-        unless r["id"].nil?
-          @runid = r["id"]
+          @run_info[:error] = true
```
Comment (rizzza) [inline lib/testrail.rb]: Looks like you have `golang` tab-spacing going on round here

Comment (sethuster) [inline lib/testrail.rb]: is that what that is?

File: lib/testrail.rb [hunk 3]
```diff
@@ -95,29 +118,35 @@ def _send_request(method, uri, data)
...
+						error = '"' + result['error'] + '"'
+					else
+						error = 'No additional error message received'
+					end
+					Log.error("[GRIDIUM::TestRail] #{response.code} - Error with request: #{error}")
 				end
-				Log.debug("[GRIDIUM::TestRail] Error with request: #{error}")
-				raise APIError.new('TestRail API returned HTTP %s (%s)' %
-					[response.code, error])
+			rescue SocketError => error
```
Comment (rizzza) [inline lib/testrail.rb]: 👍 on the specific instead of generic rescue

Comment (rizzza) [inline lib/testrail.rb]: should there be more, like a `timeout` or something similar?

Comment (rizzza) [inline lib/testrail.rb]: I guess we can handle them as they arise over time.

File: lib/testrail.rb [hunk 4]
```diff
@@ -7,79 +7,102 @@ module Gridium
...
       if Gridium.config.testrail
         @url = ENV['GRIDIUM_TR_URL'].empty? || ENV['GRIDIUM_TR_URL'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_URL'] + '/index.php?/api/v2/'
         @user = ENV['GRIDIUM_TR_USER'].empty? || ENV['GRIDIUM_TR_USER'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_USER']
         @password = ENV['GRIDIUM_TR_PW'].empty? || ENV['GRIDIUM_TR_PW'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_PW']
         @pid = ENV['GRIDIUM_TR_PID'].empty? || ENV['GRIDIUM_TR_PID'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_PID']
-				@testcase_ids = Array.new
-				@testcase_infos = Array.new
+				@config = {:pass => 1, :blocked => 2, :untested => 3, :retest => 4, :fail => 5}
+				@retry_attempts = 5
+				@time_between_retries = 3
```
Comment (rizzza) [inline lib/testrail.rb]: These two i-vars should be constants, they don't need any type of interesting initialization. Throw 'em up around line 8

File: lib/testrail.rb (hunk 1 above)
Comment (rizzza) [inline lib/testrail.rb]: Come to think of it, the test result `hash` should be a constant as well. And `freeze` it, so nobody can modify it

File: lib/testrail.rb [hunk 5]
```diff
@@ -7,79 +7,102 @@ module Gridium
...
-							)
-							sleep(0.25)
-					end
-					r = _send_request('POST', "close_run/#{@runid}", nil)
+			closed = false
+			if Gridium.config.testrail && !@run_info[:error]
+				Log.debug("[GRIDIUM::TestRail] Closing test runid: #{@run_info[:id]}\n")
+				r = _send_request('POST', "#{@url}update_run/#{@run_info[:id]}", {:case_ids => @tc_ids})
+				Log.debug("[GRIDIUM::TestRail] UPDATE RUN: #{r}")
+				sleep 0.25
```
Comment (rizzza) [inline lib/testrail.rb]: any reason why we need these sleeps?

File: lib/testrail.rb [hunk 6]
```diff
@@ -7,79 +7,102 @@ module Gridium
...
+				r = _send_request('POST', "#{@url}add_results_for_cases/#{@run_info[:id]}", {results: @tc_results})
+				Log.debug("[GRIDIUM::TestRail] ADD RESULTS: #{r}")
+				sleep 0.25
+				Log.debug("#{r.class}")
+				if r.is_a?(Hash)
+					r = _send_request('POST', "#{@url}update_run/#{@run_info[:id]}", {:name => "ER:#{@run_info[:name]}", :description => "#{@run_info[:desc]}\nThe following was returned when adding cases: #{r}"})
```
Comment (rizzza) [inline lib/testrail.rb]: not a big deal, but it's nice to have the payload in a defined variable to make the line easier to read when it is has this many characters

File: lib/testrail.rb (hunk 4 above)
Comment (sethuster) [inline lib/testrail.rb]: The reason I'm putting then in the init - is so I can override them in the test.  For some reason object.const_set doesn't work.

File: lib/testrail.rb (hunk 5 above)
Comment (sethuster) [inline lib/testrail.rb]: So we don't blast the API and get shut out...

File: lib/testrail.rb (hunk 6 above)
Comment (sethuster) [inline lib/testrail.rb]: Not sure what you mean here.  The payload is defined as the hash in the send_request function...

File: lib/testrail.rb [hunk 7]
```diff
@@ -4,120 +4,150 @@
...
+    def initialize
       if Gridium.config.testrail
         @url = ENV['GRIDIUM_TR_URL'].empty? || ENV['GRIDIUM_TR_URL'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_URL'] + '/index.php?/api/v2/'
         @user = ENV['GRIDIUM_TR_USER'].empty? || ENV['GRIDIUM_TR_USER'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_USER']
         @password = ENV['GRIDIUM_TR_PW'].empty? || ENV['GRIDIUM_TR_PW'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_PW']
         @pid = ENV['GRIDIUM_TR_PID'].empty? || ENV['GRIDIUM_TR_PID'].nil? ? ENV_ERROR : ENV['GRIDIUM_TR_PID']
-				@testcase_ids = Array.new
-				@testcase_infos = Array.new
+        @retry_attempts = 5
+        @time_between_retries = 3
```
Comment (rizzza) [inline lib/testrail.rb]: these are constants, don't belong in an `init`

File: lib/testrail.rb (hunk 6 above)
Comment (rizzza) [inline lib/testrail.rb]: it was just a comment on code readability

File: lib/testrail.rb (hunk 5 above)
Comment (rizzza) [inline lib/testrail.rb]: you could put the sleep in `_send_request` to prevent rate limiting

**Pull request merged**